### PR TITLE
CM-22718 - Support scan in batches with progress bar

### PR DESCRIPTION
**Pull request by MarshalX**
- [x] separate documents by batches
- [x] run many scans in parallel
- [x] fix races with auth token
- [x] remove hardcoded scan_id from the context
- [x] remove hardcoded issue_detected from the context
- [x] add ability to have many scan_id per CLI session
- [x] fix scan_id in async scans (generates on BE side)
- [x] fix report_url 
 - [ ] add support for scan errors (including timeouts) in batching (UI)
     - [x] collect all errors during the scan and don't show until the end
     - [x] show errors in correct --output format!
         - [ ] add JSON format support for exceptions. **Comment: not required yet**
     - [x] raise polling timeout errors 
 - [x] code refactoring. don't use inner functions if possible (for threads)
 - [x] update is_scan_failed logic by issue_detected (we have many issue_detected now)
 - [x] add --no-progress-meter flag
 - [x] disable progress bar for JSON output
 - [x] implement multi-sectioned progress bar
 - [x] integrate progress bar to all scans
 - [x] move all info logs to debug ones. only progress bar should stay
 - [x] remove halo package

Printers:
- [x] rework text printer 
    - [x] add support for aggregated scan results
    - [x] change source of scan_id
    - [x] change source of report_url
    - [x] add support for many scan_id per print (UI)
- [ ] rework json printer
    - [x] add support for aggregated scan results
    - [x] change source of scan_id
    - [x] change source of report_url
    - [ ] ❌  add support for many scan_id per print (UI). **Comment: I can't do it without changing JSON structure. I was not allowed to change it**
- [x] rework table printer
    - [x] add support for aggregated scan results
    - [x] change source of scan_id
    - [x] change source of report_url
    - [x] add support for many scan_id per print (UI)

### CODE REVIEW

File: cycode/cli/utils/progress_bar.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+    bar = get_progress_bar(hidden=False)
+    bar.start()
+
+    for bar_section in ProgressBarSection:
+        section_capacity = random.randint(500, 1000)
+        bar.set_section_length(bar_section, section_capacity)
+
+        for i in range(section_capacity):
+            time.sleep(0.01)
```
[the remark this answers is not in the document]
Comment (MarshalX) [reply cycode/cli/utils/progress_bar.py:234]: #cycode_sast_ignore_here

**Comment by MarshalX** [comment]
progress bar demo (fake progress speed):

https://github.com/cycodehq-public/cycode-cli/assets/15520314/86e298aa-c521-442f-a6d7-74c2347818e5

**Comment by MarshalX** [comment]
Errors displaying. UI synced with Maria. One row means one batch. UUID is scan_id for debug purposes: 

image

**Review by artem-fedorov — APPROVED** [review]

**Comment by MarshalX** [comment]
i place it here for future changes 
```json
// The current:
{
    "scan_id": "someUuid",
    "detections": [
        {
            "type": "slack-token",
            "detection_details": {...},
            ...
        }
     ]
}

// The new one. That will support batches
{
    "scan_results" : [
        {
            "scan_id": "someUuid",
            "report_url": "someUrl",
            // it's flexible  and we can add more field here in the future
            "detections": [
                {
                    "type": "slack-token",
                    "detection_details": {...},
                    ...
                }
            ]
        }
    ]

}

```

**Pull request merged**